HttpCookie and its tests

domaspoliakas opened the pull request:
Fixes #3921 

Not a very strict implementation of the RFCs, but I find that it largely matches how the JVM behaves. I tried copying from netty at first (as netty is also apache 2.0), but ultimately found that implementing this was more about matching the quirks of the JVM rather the RFCs, so I ended up implementing something from scratch.

I'll add some more tests, but given that this is my first PR to the project I wanted to make sure that the PR is in order. As such let me know if there's anything else needed here!

domaspoliakas committed:
HttpCookie and its tests

domaspoliakas commented:
I have signed the CLA

LeeTibbert commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,374 @@
+    )
+  }
+
+  override def equals(obj: Any): Boolean = {
```
A style point.  I recommend moving the `override` of `hashCode` below
to be either just before or just after the override of `equals()`.
That will make it easier for people reading the code who go on alert
when they see `equals` being overridden to convince themselves 
that `hashCode` is properly overridden in co-ordination.

As long as one thing is moving, I suggest moving `toString()` to
the same block so that overrides are together and easier to find.

LeeTibbert replied:
Late at night now. By the light of day I will try to post a URL
to the current "Best Practice" for an `equals` method.

It helped me last week when I had to write an `equals` for 
`fileKey`s.  Ultimately, because of some wierdness with `fileKey`,
I could not follow it exactly, but at least I knew what I was deviating from and why.

domaspoliakas replied:
Moved things around and implemented a new `equals` as per the advice you linked

LeeTibbert commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,374 @@
+  def getValue(): String = this._value
+  def getVersion(): Int = this._version
+  def hasExpired(): Boolean = this._maxAge == 0
+  override def hashCode(): Int = toString().hashCode()
```
This works and balances `equals`.

There is a recommendation that I have never quite understood as to
using `.##` instead of `hashCode` when defining a hashCode method.
I believe that `String` (and most other objects) have such.  I believe
that `.##` handles some special cases.

LeeTibbert replied:
There a are a number of `override hashCode` implementations
in the javalib code base, so it is hard to point out one which
is 'Best (or at least current) Practice'. Answering the question of
'why is that one better (or gives a better distribution)' could take
a semester course in Maths.

The one in `java.util.Arrays` gives a basic structure.  It comes from
Scala.js, where folks are knowledgeable, experienced, and super strict.

A non array body would look something like the 
example from `nio/file/attribute/PosixFileAttributeViewImpl.scala` 
```
   override def hashCode(): Int = {
      var res = 1
      res = 31 * res + deviceId.##
      res = 31 * res + inodeNumber.##
      res
    }
```
The idea is a field-by-field comparison by accumulating and shifting the intermediate results.  If the fields can be null, one must handle that (long discussion).

If I live a Good Life, Someday I may be saved from having to write 
equals/hashcode pairs.

domaspoliakas replied:
The JVM seems to include the name, value, domain and path into the hashcode, so I changed mine to do so as well, using the formula you gave there.

LeeTibbert commented on unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/HttpCookieTest.scala:
```diff
@@ -0,0 +1,73 @@
+    val cookie = out.get(0)
+    assertEquals("potato", cookie.getName())
+    assertEquals("tomato", cookie.getValue())
+    assertEquals(null, cookie.getComment())
```
A comment based on my personal style. Authors have wide latitude
to follow almost any correct style.

When writing tests, I like to take a few moments to think about
how a person running the test might figure out what went wrong and
where.

When there is only one assert in a `Test`, that is not a problem.
When there are 20 or 40 or so (some of the regex code), it can
mean many hours of tedious bi-section search.    

`assertEquals` and many of its kin have a three argument form,
where the first argument is a message.  It takes a developer 
minutes of time to place a message, but it can save a debugger hours.
Usually when debugging, time is of the essence: thing are not good and are probably going to get worse.

Here one can probably figure out the failure point if one of the 
first two fails first two.  I suspect that line 20 and kin would 
cause some puzzlement.

A quick search shows `unit-tests/shared/src/test/require-scala3-jdk12/org/scalanative/testsuite/javalib/util/stream/CollectorsTestOnJDK12.scala` following that practice.

Now meaningful messages are an ideal.  In some cases where I 
have had to figure out which of 5 or more assertions (especially
the dreaded `assertTrue` & `assertFalse`, which do not give values)
are failing and have less than zero time to do it, I fall back to 
numbering the assertions "a1", "a2", etc.  This gives me a brainless
way to distinguish them. 

Yes, I hang my head that I should be taking the time to figure out a better message.  That means figuring out what all sorts of quirky code
is supposed to be doing, where that code has nothing to do with my
presenting problem.

The downside is that when I add or delete assertions, I have to do the
tedious job of renumbering, so that later maintainers can
follow the intent.   Paying for my transgressions.

domaspoliakas replied:
> It takes a developer minutes of time to place a message, but it can save a debugger hours.

Lovely piece of advice :) Using `CollectorsTestOnJDK12` as a template I tried to add some similar clues to the assertions, let me know if that's what you're looking for here.

LeeTibbert commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,374 @@
+
+  }
+
+  private[HttpCookie] val Discard = "Discard"
```
Please pardon my lack of understanding.
What does specifying the scope `[HttpCookie]` add here?
How does it differ from `private val`.

I thought that companion classes and corresponding object
could reference each other's private variables.  Yes, the 
class must either import the symbols of the companion Object
or do someething like `HttpCookie.Discard`.

I just checked and one can have `final` vals of type `String`.  I know
that adding `final` helps the SN compiler treat numeric vals as a
compile time constant.  I do not know if that works with String.
Perhaps someone with better knowledge of Scala Native internals
can chime in.

Anyway I think `final` will not hurt.

LeeTibbert commented on unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/HttpCookieTest.scala:
```diff
@@ -0,0 +1,73 @@
+    assert(!HttpCookie.domainMatches("website.com", "awebsite.com"))
+  }
+
+}
```
One can spend eternities one does not have writing tests
for failure cases.  Users are _so_ good at generating just
the failure case not detected by tests.  Such is life.

The code goes to some length to detect`IllegalArgumentException("Illegal cookie name")`.

Did I miss a Test for that?  If not, you might want to consider adding it.

(is the "Illegal" in the message from JVM? If not, could it be changed
to something which does not keep me up all day wondering if the Sheriff's Deputy is coming for me because I typo-ed a name)

domaspoliakas replied:
> is the "Illegal" in the message from JVM?

Yes, I tried to mimic the behavior of the JVM; at least to a reasonable degree. Added tests for that now as well

LeeTibbert commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,374 @@
+  }
+
+}
+
```
There may need to be a javadoc/scaladoc comment here
so that the documentation build picks this up as a supported class.

@ekrich  What is the current process for a newly supported class
                to get picked up by the documentation?

                 I blinked and the process changed.
                 Thanks.

LeeTibbert commented:
domaspoliakas 

Thank you for this contribution. I hope it is the first of many.

Overall, it looks very good. It is nice to see the corresponding & careful Tests.

I did my traditional "What documentation impacts does this PR have?" and
found out I/we need a short tutorial on current practice from @ekrich.
I think the required change is to add a javadoc/scaladoc (i.e. ``/**  HttpCookie*/`) comment before
the new class. I may be way off base here.  Guess this PR gives me the opportunity to
learn something contemporary.

Enough for tonight, I will check the log files by the light of day to see what is failing & try to
puzzle out why.

LeeTibbert commented:
To allow my mind to rest and focus on my current project, I did a spot test of perhaps 3 failing
tests.  They were all the same:
```
    while (i < unprocessed.size) {
[error]    |                 ^^^^^^^^^^^^^^^^
[error]    |         method size in trait Collection must be called with () argument
```

Hope that helps.

LeeTibbert commented:
As promised, FYI:
 https://alvinalexander.com/scala/how-to-define-equals-hashcode-methods-in-scala-object-equality/

I found this article helpful. In it Alvin references at least two of his original sources.  Alvin is careful
and informed, not "just some joe off the Internet". 

An almost identical descriptions occur on `scala-lang.org`: my guess is that the two are related.
Probably both his work.

If I understand `HttpCookie` correctly, it can be extended which means that it can not be marked `final`.
That means that one must figure out a proper `canEquals` (case? I am writing from memory).

Building up enough recent models of the current recommended practice will make it easier for people
to find the models.

LeeTibbert commented:
Any idea what the chance of a hash collision is from "concatenationOfAllFields".##?  I sure do not
know and have no idea how to figure that math.  Likely to be relatively expensive at runtime but 
good, easy devo bootstrapping.

domaspoliakas committed:
HttpCookie: more tests, more docs, some bugfixes

domaspoliakas commented:
Addressed some of your comments wrt equals, hashcode and tests.

For hashcode I used `name`, `value`, `domain` and `path` as that seems to be what the JVM is doing as well.

Also it turns out the JVM keeps track of the time when the object was created for the purpose of evaluating `hasExpired()`, which seems crazy to me but nevertheless I have also replicated here now.

domaspoliakas committed:
Formatting

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+  def getValue(): String = this._value
+  def getVersion(): Int = this._version
+  def hasExpired(): Boolean = {
+    _maxAge >= 0 && ((System
```
JVM would return true if `maxAge == 0`, so we should start with 2 quick range checks before starting getting current time

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+  def getComment(): String = this._comment
+  def getCommentURL(): String = this._commentURL
+  def getDiscard(): Boolean = this._discard
+  def getDomain(): String = this._domain
```
JVM stores domain as lowercase 

```scala
                                                                                                                                                                                                                                                                                                                                                
scala> new java.net.HttpCookie("foo", "bar")
val res0: java.net.HttpCookie = foo="bar"
                                                                                                                                                                                                                                                                                                                                                
scala> res0.getDomain
val res1: String = null
                                                                                                                                                                                                                                                                                                                                                
scala> res0.setDomain("MyDomain")
                                                                                                                                                                                                                                                                                                                                                
scala> res0.getDomain
val res2: String = mydomain
```

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+  def setPortlist(ports: String): Unit = this._portlist = ports
+  def setSecure(flag: Boolean): Unit = this._secure = flag
+  def setValue(newValue: String): Unit = this._value = newValue
+  def setVersion(v: Int): Unit = this._version = v
```
JVM requires version to be either 0 or 1 
`Throws: IllegalArgumentException - if v is neither 0 nor 1`

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+        if (char == separator && !inQuotes) {
+          res.add(in.substring(begin, i))
+          begin = i + 1
+        } else if (char == '\\' && inQuotes && !escaping) {
+          escaping = true
+        } else if (char == '"' && inQuotes && !escaping) {
+          inQuotes = false
+        } else if (char == '"' && !inQuotes) {
+          inQuotes = true
+        }
```
How about refactoring `inQuotes && !escaping` into a common condition so we don't need to check it twice? 
```suggestion
        } else if (inQuotes){
         if(!escaping) char match {
           case '\\' =>  escaping = true
           case '"'  =>  inQuotes = false
           case _ => ()
        }
        else if (char == '"') inQuotes = true
      
```

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+            prop.substring(0, sep).trim()
+          }
+
+        if (propName.regionMatches(true, 0, Discard, 0, Discard.length())) {
```
How about `equalsIgnoreCase` instead? Again, you're comparing the whole string. Alternatively you can convert the `prop.toLowerCase` and then make a simple match - would be way more readable.

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+            )) {
+          try {
+            val value = java.lang.Integer.valueOf(propValue)
+            if (value != 0 && value != 1) {
+              throw new IllegalArgumentException(
+                "cookie version should be 0 or 1"
+              )
+            }
```
This check should be moved to `setVersions

WojciechMazur commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+  override def equals(obj: Any): Boolean =
+    obj match {
+      case other: HttpCookie =>
+        super.equals(other) &&
```
Also the javadoc states that `The result is true only if two cookies come from same domain (case-insensitive), have same name (case-insensitive), and have same path (case-sensitive).
` So there is no need to check all the fields

WojciechMazur reviewed (changes requested):
Thank you for the contribution! It seams to not be passing the CI - it's failing due to incompatibilities with the JVM semantics - we ran javalib tests on both JVM (first) and then using Native. Because it fails on the JVM it means that tests are incorrect and don't follow JVM semantics. You can run these locally using `testsJVM3/test`

domaspoliakas commented on javalib/src/main/scala/java/net/HttpCookie.scala:
```diff
@@ -0,0 +1,413 @@
+    else {
+
+      val nameless =
+        if (header.regionMatches(false, 0, "set-cookie2:", 0, 12)) {
+          header.substring(12)
+        } else if (header.regionMatches(false, 0, "set-cookie:", 0, 11))
+          header.substring(11)
```
`startsWith` is case-sensitive, which is why I use regionMatches here

domaspoliakas committed:
HttpCookie - various refactors to address PR comments

domaspoliakas committed:
HttpCookie WIP to match JVM semantics

domaspoliakas commented:
Addressed the PR comments now. There's one mismatch between the JVM behavior and my implementation that I'm still trying to track down though, I'll keep working on it in the meantime.

lqhuang commented:
Closable since there's an out-of-tree implementation available now. You can check it out here <https://github.com/lqhuang/scala-native-http/pull/1>

Huge thanks to Domas Poliakas for his work!

LeeTibbert commented:
@lqhuang @domaspoliakas 

Closing as requested.  May  the out-of-tree implementation prosper.

If/when I can steal/create some cycles, I may look  at the SN documentation to
see how we can make it easier for general application developers to find.
Probably near where we describe 'java.time'  (@ekrich, anything of the top of your head?
I do not want to distract you.)

It is satisfying when  progress actually happens. Thank you (and rishi-jat).
